File: datastrutures/List.java
class ElementList {
    ElementList(int value){
        this.value = value;
    }
    public int value;
    public ElementList next=null;
}

public class List {
    ElementList head = null; //first element on list
    ElementList tail = null; //last element on list

    int length=0;


    public void add(int value) {
        ElementList a1 = new ElementList(value);
        if (isEmpty()){
            head=a1;
            tail=a1;
        }
        else{
            tail.next = a1;
            tail=a1;
        }
        length++;
    }

    public void deleteAt(int index)
    {
        if(isEmpty()){//what if list is empty
            System.out.println("List is empty!!");
            return;
        }

        if(index==0 && length>1){ //what if i delete the first element and the size is greater than 1
            head = head.next;
        }else if(index==0 && length==1){//what if i delete the first item and i have 1 element
            head=null;
            tail=null;
        } else if(index==length-1){ //what if i delete last element
            ElementList a = head;
            for (int i = 0; i < length-1; i++) {
                a = a.next;
            }
            tail=a;
        } else if(length==3 && index==1){ //what if i remove element from middle while i have 3 elements
            head.next = tail;
        } else if(length>3 && index>=0){ //what if i remove element from middle, and i have more than 3 elements - normal case
            ElementList previous = null;
            ElementList toDelete = head;

            for (int i = 0; i < index && toDelete != null; i++) {
                previous = toDelete;
                toDelete = toDelete.next;
            }

            if (toDelete == null) {
                System.out.println("There is no item with this index");
                return;
            }

            previous.next = toDelete.next;


        } else if(index>length+1 || index<0){
            System.out.println("There is no item with this index");
            return;
        }
        length--;
    }

    public Integer getValue(int index) {
        if (index >= length || index<0 ) {
            System.out.println("Given value doesn't exist on the list");
            return null;
        }
        else
        {
            ElementList a = head;
            for (int i = 0; i < index; i++)
            {
                a = a.next;
            }

            return a.value;
        }
    }

    public int getIndex(int value) {
        int index = 0;
            while (index <= length) {
                if (index >= length) {
                    System.out.println("Given index doesn't exist on the list");
                    return -1;
                }else{
                ElementList a = head;

                for (int i = 0; i < length; i++) {
                    if (getValue(i) == value)
                        return index;
                    else {
                        a = a.next;
                        index++;
                    }
                }
            }
        }
        return index;
    }
    public boolean isEmpty()
    {
        if(head==null)
            //return head==null;
            return true;
        else return false;
    }



    public void print() {
        ElementList a = head;
        for (int i = 0; i < length; i++)
        {
            System.out.println(a.value);
            a = a.next;
        }
    }
}

File: datastrutures/list-cpp/list.cpp
#include <iostream>

class Element{
public:
    Element* next = nullptr;
    int value;

    Element(long int value){
        this->value = value;
        next = nullptr;
    }
};

class List{
public:
    Element* head = nullptr;
    Element* tail = nullptr;
    int length=0;


    void add_after_element(int value, int index){
        if (index < 0 || index >= length) {
            // invalid index
            std::cout << "Invalid index" << std::endl;
            return;
        }

        Element* temp = new Element(value);
        if (isEmpty()) {
            head = temp;
            tail = temp;
        } else {
            Element* after = get(index);
            if (after == tail) {
                tail->next = temp;
                tail = temp;
            } else {
                // Inserting in the middle
                temp->next = after->next;
                after->next = temp;
            }
        }

        length++;
    }

    void append(int value){
        Element* temp = new Element(value);
        if(isEmpty()){
            head = temp;
            tail = temp;
        }
        else{
            tail->next=temp;
            tail=temp;
        }
        length++;
    }

    void add_to_start(int value){
        Element* temp = new Element(value);
        if(isEmpty()){
            head = temp;
            tail = temp;
        }
        else{
            temp->next=head;
            head=temp;
        }
        length++;
    }

    void deleteAt(int index){
        Element* temp;
        if(isEmpty()){//what if list is empty
            std::cout<<"list is empty!!";
            return;
        }

        if(index==0 && length>1){ //what if I remove the first element and the size is greater than 1
            head = head->next;
        }else if(index==0 && length==1){//what if I remove the first element and I have 1 element
            head= nullptr;
            tail=nullptr;
        } else if(index==length-1){ //what if i delete the last item
            temp = head;
            for (int i = 0; i < length-1; i++) {
                temp = temp->next;
            }
            tail=temp;
        } else if(length==3 && index==1){ //what if I remove an element from the middle while I have 3 elements
            head->next = tail;
        } else if(length>3 && index < length-1){ //what if I remove an element from the middle, and I have more than 3 elements - normal case
            Element* previous;
            Element* toDelete = head;
            for (int i = 0; i < index; i++) {
                previous = toDelete;
                toDelete = toDelete->next;
            }
            previous->next= toDelete->next;
        }
        else std::cout<<"There's no element with given index"<<std::endl;
        length--;
    }

    Element* get(int index){
        if(index >= length){
            std::cout<<"Does not exists on the list";
            return NULL;
        }
        else{
            Element* temp = head;
            for(int i =0; i<index; i++){
                temp = temp->next;
            }
            return temp;
        }
    }


    bool isEmpty() {
        if (head == nullptr)
            return true;
        else { return false; }
    }

    void print() {
        Element* temp = head;
        for(int i=0; i < length; i++){
            std::cout << temp->value << std::endl;
            temp = temp->next;
        }
    }
};